command.go

```diff
@@ -622,6 +637,137 @@ func (c *commandContext) FinalizeCurrentMode() error {
 	return nil
 }
 
+func (c *commandContext) FitInserting(axes [6]bool) error {
+	if c.selectMode != selectModeInsert {
+		return errors.New("not in insert mode")
+	}
+	it, err := c.editor.pp.Vec3Iterator()
+	if err != nil {
+		return err
+	}
+	itSubOrig, err := c.editor.ppSub.Vec3Iterator()
+	if err != nil {
+		return err
+	}
+	trans := cursorsToTrans(c.selected)
+	itSub := &transformedVec3RandomAccessor{
+		Vec3RandomAccessor: itSubOrig,
+		trans:              trans,
+	}
+	minMain, maxMain, err := pc.MinMaxVec3(it)
+	if err != nil {
+		return err
+	}
+	minSub, maxSub, err := pc.MinMaxVec3(itSub)
+	if err != nil {
+		return err
+	}
+
+	const (
+		matchRange        = 0.5
+		regionPadding     = 1.0
+		maxBasePoints     = 60000
+		maxTargetPoints   = 10000
+		minSampleRatio    = 0.01
+		gradientWeight    = 0.3
+		gradientPosThresh = 0.001
+		gradientRotThresh = 0.002
+		maxIteration      = 50
+	)
+
+	is := rectIntersection(
+		rect{minMain, maxMain},
+		rect{minSub, maxSub},
+	)
+	is.min = is.min.Sub(mat.Vec3{regionPadding, regionPadding, regionPadding})
+	is.max = is.max.Add(mat.Vec3{regionPadding, regionPadding, regionPadding})
+	if !is.IsValid() {
+		return errors.New("no intersection")
+	}
+	center := is.min.Add(is.max).Mul(0.5)
+
+	sample := func(ra pc.Vec3RandomAccessor, isIn func(mat.Vec3) bool, nMax int) (pc.Vec3Slice, float32) {
+		var cnt int
+		for i := 0; i < ra.Len(); i++ {
+			if isIn(ra.Vec3At(i)) {
+				cnt++
+			}
+		}
+		ratio := float32(nMax) / float32(cnt)
+		out := make(pc.Vec3Slice, 0, nMax)
+		for i := 0; i < ra.Len(); i++ {
+			if p := ra.Vec3At(i); isIn(p) {
+				if rand.Float32() > ratio {
+					continue
+				}
+				out = append(out, p.Sub(center))
+				if len(out) >= nMax {
+					break
+				}
+			}
+		}
+		return out, ratio
+	}
+
+	base, ratioBase := sample(it, is.IsInside, maxBasePoints)
+	if ratioBase < minSampleRatio {
+		return errors.New("too many base points")
```

Comment:
Is it `too many`? `ratioBase` being small means that the number of point inside the intersection is much larger than`maxBasePoints`. It seems like the error message should be `too few` then.

Reply:
Sampling ratio is determined by `maximum_points / given_points` , so the sampling ratio will be smaller when the number of the points are large.

Reply:
`minSampleRatio` is to avoid losing features by removing too many amount of points.
I think there are basically no way to do with too large pointclouds on browser. (excepting optimization of the program)

In the perspective of the user, sampling is invisible and just internally done. From outside, the problem is that the given pointcloud is too large.

Reply:
I see. So when this error occurs, the only thing a user could do, would be to reduce the overlap between the base pointcloud and the inserted one. Is that correct?